src/umode.rs

```diff
@@ -429,7 +444,58 @@ impl UmodeTask {
         res.map_err(Error::Request)
     }
 
-    fn handle_ecall(&mut self) -> ControlFlow<Result<OpResult, ExecError>> {
+    fn handle_cdi_op(
+        attestation: Option<&AttestationManager<sha2::Sha384>>,
+        cdi_sel: CdiSel,
+        cdi_op: CdiOp,
+    ) -> Result<(), ExecError> {
+        if let Some(attmgr) = attestation {
+            let cdi = match cdi_sel {
+                CdiSel::AttestationCurrent => Ok(attmgr.attestation_current_cdi()),
+                _ => Err(ExecError::UnexpectedCdiOp(cdi_sel, cdi_op)),
+            }?;
+            match cdi_op {
+                CdiOp::Id {
+                    idout_addr,
+                    idout_len,
+                } => {
+                    let id = cdi.id().map_err(ExecError::Cdi)?;
+                    if idout_len as usize != id.len() {
+                        return Err(ExecError::BufferSize(idout_len, id.len() as u64));
+                    }
+                    HypMap::copy_to_umode(RawAddr::supervisor_virt(idout_addr), &id)
+                        .map_err(ExecError::UmodeAccess)
+                }
+                CdiOp::Sign {
+                    msg_addr,
+                    msg_len,
+                    signout_addr,
+                    signout_len,
+                } => {
+                    let mut msg_buf = [0u8; CDIOP_SIGN_MAXMSG];
+                    if msg_len > CDIOP_SIGN_MAXMSG as u64 {
+                        return Err(ExecError::BufferSize(msg_len, CDIOP_SIGN_MAXMSG as u64));
+                    }
+                    let msg = &mut msg_buf[0..msg_len as usize];
+                    HypMap::copy_from_umode(msg, RawAddr::supervisor_virt(msg_addr))
+                        .map_err(ExecError::UmodeAccess)?;
+                    let signature = cdi.sign(msg).to_bytes();
+                    if signout_len as usize != signature.len() {
+                        return Err(ExecError::BufferSize(signout_len, signature.len() as u64));
+                    }
+                    HypMap::copy_to_umode(RawAddr::supervisor_virt(signout_addr), &signature)
+                        .map_err(ExecError::UmodeAccess)
```

Comment:
i'm not thrilled about all the copying here, but i suppose that can be saved for a future optimization.

Reply:
Yes, avoiding copy would either require as to make special slices that are active under SUM enabled, or change the interface of the signer (or rice).